server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java

```diff
@@ -144,9 +156,26 @@ public ActionRequestValidationException validate() {
         if (settings == null) {
             validationException = addValidationError("settings is null", validationException);
         }
+        if (isMetadataTooBig(userMetadata)) {
+            validationException = addValidationError("metadata must be smaller than 1024 bytes", validationException);
+        }
         return validationException;
     }
 
+    private static boolean isMetadataTooBig(Map<String, Object> userMetadata) {
+        if (userMetadata == null) {
+            return false;
+        }
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            builder.value(userMetadata);
+            int size = BytesReference.bytes(builder).length();
+            return size > MAXIMUM_METADATA_BYTES;
```

Comment:
Maybe we could just return the size of the meta here and then do the check upstream? That way we could add the serialized size of the metadata to the exception which might be helpful when debugging things in the future?

Reply:
I've made this change, it's a good suggestion.